jscomp/ml/typecore.ml

```diff
@@ -1864,36 +1864,44 @@ and type_expect ?in_function ?recarg env sexp ty_expected =
          type_expect_ ?in_function ?recarg env sexp ty_expected
       )
   in
-  let () =
-    let rec extractPromise t =
-      match t.desc with
-      | Tconstr (Pdot (Pdot (Pident {name = "Js"}, "Promise", _), "t", _), [t1], _)
-        ->
+  checkTypeInvariant exp;
+  Cmt_format.set_saved_types
+  (Cmt_format.Partial_expression exp :: previous_saved_types);
+  exp
+
+(* NOTE: the type invariant check should have no side effects and be efficient *)
+and checkTypeInvariant exp : unit =
+  let rec extractPromise t =
+    match t.desc with
+    | Tconstr
+        (Pdot (Pdot (Pident { name = "Js" }, "Promise", _), "t", _), [ t1 ], _)
```

Comment:
This is not needed?

Reply:
could this be made opt-in? If this is a built in type, I remember there're some faster check

Reply:
I would remove `Js.Promise.t` from `js_promise.ml` but that's a breaking change.
So as long as you can use it explicitly, it's easy to unintentionally disable the check.
Probably something to remove a little later when the feature is less experimental.